### DISCUSSION TITLE
Kademlia Events

### DISCUSSION THREAD

**Discussion by thewoodfish** [libp2p/rust-libp2p#5339] [Q&A]
In `libp2p-rust`, events are generated. For kademlia, does `kad::Event::InboundRequest` generate event after libp2p kad has handled the incoming request or is the event generated after the request has been received and the user is means to explicitly handle the request? e.g the `PutRecord` event. Is it generated after Kademlia has saved the incoming request and then publishes an event or it generates the event after the request is received and expects the implementer to handle the request explicitly with e.g `kad::Behaviour::put_record()`

**Comment by dariusc93** [comment]
Hey @thewoodfish. libp2p-kad will yield an `Event::InboundRequest` for handled incoming request, though for the user to handle the request explicitly, they would have to set `StoreInserts::FilterBoth` so you can manually insert the record into the store, otherwise it would be handled automatically internally as `StoreInserts::Unfiltered` is default.

**Reply by thewoodfish** [reply]
`though for the user to handle the request explicitly, they would have to set StoreInserts::FilterBoth so you can manually insert the record into the store`... The explicit handling would be done in the event handling section because of the data exposed yeah?
